docs/design_proposals/debug-events.md

```diff
@@ -0,0 +1,221 @@
+When a new debuggable container is created, Skaffold will
+attempt to forward the remote debug port, which should result in a port-forward event, and thus
+spare the IDEs from having to actively watch the cluster. 
+
+But relying strictly on port-forward events carries some risk.  Although the likelihood of 
+a port-name clash is low, the IDE may require additional configuration information to establish
+a debugging connection.  For example, there may be no indication of the underlying project’s
+location in the file-system, nor the relative path in the container. 
+
+Although such information is published by `skaffold debug` via pod annotations, obtaining this
+information would require that the IDEs actively track new pods and containers.  This tracking would
+duplicate functionality already being performed by Skaffold, and also requiring this functionality to be
+re-implemented in each IDE.
+
+This design proposal recommends that Skaffold instead also watch and issue events when debuggable
+containers are started and terminated.
+
+----
+## Design
+
+This design proposes that Skaffold issue events to inform the IDEs of the following conditions:
+
+- The appearance of a container that is debuggable.  This event will provide:
+    - the namespace
+    - the pod ID
+    - the container ID
+    - the contents of the `debug.cloud.google.com/config` configuration block for the container
+      (a JSON object) including (! indicates new items)
+         - the language runtime type (`go`, `jvm`, `nodejs`, `python`)
+         - ! the image name (which should allow cross-referencing with the `skaffold.yaml`)
+         - the port(s)
+         - ! the container image's working directory
+         - ! (optional) the container image's _remote root_ for source file resolving 
+
+- The disappearance of a container that was debuggable.  This event will provide:
+    - the namespace
+    - the pod ID
+    - the container ID
+    - the contents of the `debug.cloud.google.com/config` configuration block for the container
+      (a JSON object) including (! indicates new items)
+         - the language runtime type (`go`, `jvm`, `nodejs`, `python`)
+         - ! the corresponding artifact's `name` (which should allow cross-referencing with the `skaffold.yaml`)
+         - the port(s)
+         - ! the container image's working directory
+         - ! (optional) the container image's _remote root_ for source file resolving 
+         
+Skaffold will also maintain state and report on the set of debuggable containers.
+
+With this information, the IDE should be able to:
+
+- Identify the corresponding artifact within the `skaffold.yaml`
+- Identify the corresponding artifact from a port-forward event by correlating the
+  remote port with the debug port from the previously-raised debug event
+- Establish new debug launch configurations with the forwarded port,
+  and the local source location, and any other necessary information
```

Comment:
As the local source location is needed by both IJ and VSCode extensions, can we have a way to store this info in the skaffold.yaml configuration?

Reply:
With this, we don't need to create a 'debug' node in the vscode launch configuration.

Reply:
The local source root is unfortunately not really clear — certainly the `debug` portion has no more information than the `skaffold.yaml`'s artifact `context` directory.  Consider a source layout like:
```
path/to/my/artifact/
├── Dockerfile
└── some
    └── deep
        └── directory
            └── RandomPython.py
            └── RandomPython.py
            └── RandomPython.py
            └── RandomPython.py
└── another
    └── deep
        └── directory
            └── RandomPython.py
            └── RandomPython.py
            └── RandomPython.py
            └── RandomPython.py
```

where the `Dockerfile` is:
```
COPY some/deep /random/place/in/image
COPY another/deep/directory /another/place/in/image
WORKDIR /a/different/location
```

I suspect we'll find it worthwhile to develop a little Go utility to guess at possible source roots, (if something like this doesn't already exist?).